### PR TITLE
tools: topology: Add common Codec Adapter pipeline tplg

### PR DESCRIPTION
**Pull request by johnylin76**
pipe-codec-adapter-playback.m4 and pipe-codec-adapter-capture.m4 is implemented as a common pipeline module of a playback/capture codec w/ Codec Adapter API.

For codec developers, a upper-layer tplg file should be provided for the specified codec with the information of control bytes, schedule core, and byte-control names.

### CODE REVIEW

File: tools/topology/sof/pipe-codec-adapter-playback.m4 [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+include(`pipeline.m4')
+include(`codec_adapter.m4')
+include(`bytecontrol.m4')
+
+#
+# Controls
+#
+
+# For codec developers, please define the following control bytes of setup config and runtime params
+# and the corresponding max bytenum on your own.
```
Comment (lgirdwood) [inline tools/topology/sof/pipe-codec-adapter-playback.m4:21]: Can we have more context in the comment here e.g. what tools would be used, any example. just basic info to help put folks on teh right track. btw, should these headers be all 0s as "empty" controls ?

**Review by lgirdwood — COMMENTED** [review]
LGTM, I think it just needs a little more context so codec devs know how to generate data.
@ranj063 fyi for topology 2.0 updates.

File: tools/topology/sof/pipe-codec-adapter-playback.m4 [hunk 2]
```diff
@@ -0,0 +1,112 @@
+# Low Latency Passthrough Pipeline and PCM
+#
```
Comment (dbaluta) [inline tools/topology/sof/pipe-codec-adapter-playback.m4:2]: @johnylin76 I don't understand exactly WHY this change is needed? 

pipe-processing-playback.m4 file has the exact purpose of a common codec adapter playback pipeline topology that others can use in their m4 files to create a topology file.

Comment (cujomalainey) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:2]: this does feel very similar to pipe-processing, @johnylin76 can you share the intent of this change?

Comment (johnylin76) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:2]: The intent of this change is to create a template of pipi-playback m4 for codec adapter codecs. For more and more codecs in the future, it can minimize the effort and avoid errors on copy/paste.
Here is another codec we are developing: https://github.com/thesofproject/sof/pull/3718/commits/f960b11ffbd4ecb0c360bf1ac6978b88788ee23d#diff-e5866322398b5edd5d9432288f69de57626645f898caf11d02e3923787dd7eaa

File: tools/topology/sof/pipe-codec-adapter-playback.m4 (hunk 1 above)
Comment (johnylin76) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:21]: I added some contexts on it, do you think there is any information needed to append?

> btw, should these headers be all 0s as "empty" controls ?

I would prefer a valid but dummy default here, "empty" also sounds good to me.

File: tools/topology/sof/pipe-codec-adapter-playback.m4 (hunk 2 above)
Comment (cujomalainey) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:2]: Isnt' that what https://github.com/thesofproject/sof/blob/master/tools/topology/sof/pipe-processing-playback.m4 is though? Sorry still unclear on this change.

Comment (johnylin76) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:2]: Could a developer create a pipe-xxx-playback.m4 for a new codec by including pipe-processing-playback.m4 and then re-declare its control bytes of setup config and runtime param? or change the byte control port name? I don't know how to do it but it can be achieved then this change is not necessary.

File: tools/topology/sof/pipe-codec-adapter-playback.m4 (hunk 1 above)
Comment (lgirdwood) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:21]: ok - as long as this is in the comments so it's easy to follow.

File: tools/topology/sof/pipe-codec-adapter-playback.m4 (hunk 2 above)
Comment (lgirdwood) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:2]: I see this is useful change, as a higher level abstraction on the codec adapter. Fwiw, this could be one of the last complex topologies using M4 as the topology 2.0 work is maturing.

**Review by lgirdwood — APPROVED** [review]

File: tools/topology/sof/pipe-codec-adapter-playback.m4 (hunk 2 above)
Comment (cujomalainey) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:2]: Ah, yes, hopefully tplg2 will have a useful mechanism to pass in byte, and control sizes

File: tools/topology/sof/pipe-codec-adapter-capture.m4 [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+#    - [4:7]: reserved 0s
+#    - [8:11]: sample rate, e.g. 48000
+#    - [12:15]: sample width in bits, e.g. 32
+#    - [16:19]: channels, e.g. 2
+# - (optional) 12+ bytes codec_param: codec TLV parameters container, for more details please refer
+#                                     struct codec_param under audio/codec_adapter/codec/generic.h
+#    - [0:3]: param ID
+#    - [4:7]: size in bytes (ID + size + data)
+#    - [8:n-1]: data[], the param data
+ifdef(`CA_SETUP_CONTROLBYTES',`', `define(`CA_SETUP_CONTROLBYTES',
```
Comment (cujomalainey) [inline tools/topology/sof/pipe-codec-adapter-capture.m4:44]: random question, how does this work if two adapter pipelines are included? shouldnt we just implicitly unset this and reset it?

Comment (johnylin76) [reply tools/topology/sof/pipe-codec-adapter-capture.m4:44]: good question. I think we should undefine them in the end of pipe-codec-adapter-[capture|playback].m4 to void this issue.

Comment (cujomalainey) [reply tools/topology/sof/pipe-codec-adapter-capture.m4:44]: given you have now unset, isn't ifdef unneeded?

Comment (johnylin76) [reply tools/topology/sof/pipe-codec-adapter-capture.m4:44]: These define variables are used for a codec pipeline to overwrite. For example, pipe-processing-playback.m4 defines its own CA_SETUP_CONTROLBYES, and CA_RUNTIME_CONTROLBYTES, and others.
"ifdef" here is to set a default value.
And in order not to re-define variables in the cases of multiple codecs, I chose to "undefine" in the end of pipe-codec-adapter-playback.m4

Comment (cujomalainey) [reply tools/topology/sof/pipe-codec-adapter-capture.m4:44]: Yes, im just suggesting we can drop the `if` since we are unsetting above and just put a `define` here since it will always be undefined.

File: tools/topology/sof/pipe-codec-adapter-playback.m4 [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+ifdef(`CA_RUNTIME_CONTROLBYTES_MAX',`', `define(`CA_RUNTIME_CONTROLBYTES_MAX', 157)')
+
+# For codec developers, rename bytes control names if necessary.
+ifdef(`CA_SETUP_CONTROLBYTES_NAME',`',
+	`define(`CA_SETUP_CONTROLBYTES_NAME', `CA Setup Config')')
+ifdef(`CA_RUNTIME_CONTROLBYTES_NAME',`',
+	`define(`CA_RUNTIME_CONTROLBYTES_NAME', `CA Runtime Params')')
+
+define(CA_SETUP_CONFIG, concat(`ca_setup_config_', PIPELINE_ID))
+define(CA_SETUP_CONTROLBYTES_NAME_PIPE, concat(CA_SETUP_CONTROLBYTES_NAME, PIPELINE_ID))
```
Comment (johnylin76) [inline tools/topology/sof/pipe-codec-adapter-playback.m4:73]: I combined the control bytes name with pipeline ID when I saw most of other component topologies have done this. This could prevent the name is repeated if there are more pipelines adopt the codec.

File: tools/topology/sof/pipe-codec-adapter-capture.m4 (hunk 3 above)
Comment (johnylin76) [reply tools/topology/sof/pipe-codec-adapter-capture.m4:44]: Sorry could you deliberate more on "since we are unsetting above"? I still didn't get what is unsetting?

File: tools/topology/sof/pipe-processing-playback.m4 [hunk 5]
```diff
@@ -38,74 +21,26 @@ CONTROLBYTES_PRIV(PP_SETUP_CONFIG,
...
-        CONTROLBYTES_EXTOPS(void, 258, 258),
-        , , ,
-        CONTROLBYTES_MAX(void, 157),
-        ,
-        PP_RUNTIME_PARAMS)
-
-#
-# Components and Buffers
-#
+define(`CA_SCHEDULE_CORE', 1)
```
Comment (cujomalainey) [inline tools/topology/sof/pipe-processing-playback.m4:37]: Should we add pipeline ID to this?

File: tools/topology/sof/pipe-codec-adapter-capture.m4 (hunk 3 above)
Comment (cujomalainey) [reply tools/topology/sof/pipe-codec-adapter-capture.m4:44]: ah i misread the code on my last pass, I thought you were doing this
```
unset
ifdef
```
which means that our namespace should always be cleared. I think it might be better assuming unsetting an undefined does not result in an error

File: tools/topology/sof/pipe-processing-playback.m4 (hunk 5 above)
Comment (johnylin76) [reply tools/topology/sof/pipe-processing-playback.m4:37]: I don't think so, pipeline ID is brought by the upper layer.
I observed that in previous pipe-processing-playback.m4, the schedule core of codec adapter (PP_CORE) is set to 1 as default which is different than others (core=0).

**Review by cujomalainey — APPROVED** [review]
offline discussion solved issues

**Comment by cujomalainey** [comment]
Ignoring travis failures since topology should not be part of test suite on failing devices
